### PR TITLE
fix(aws-batch): Support omitting ComputeEnvironment security groups so that they can be specified in Launch Template

### LINKED ISSUE #21577: (aws-batch): (Compute environments cannot be created with launch templates specifying network interface)
### Describe the bug

Many HPC applications require low latency, and so it's desirable to use launch templates to configure EC2 instances with Elastic Fabric Adapters.   This currently fails at deployment time.

### Expected Behavior

Should be able to configure a Compute Environment with no security groups, using network interfaces in the Launch Template.

### Current Behavior

L2 Construct always creates a SecurityGroupIds property in the compute environment, and so the deployment fails with:

```
Failed resources:
batch-stack | 09:35:08 | CREATE_FAILED        | AWS::Batch::ComputeEnvironment        | EFABatch (EFABatchXXXXXXX) Resource handler returned message: "Error executing request, Exception : Either compute environment Security Groups or Network Interfaces in Launch template are exclusively allowed, RequestId: nnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn (Service: Batch, Status Code: 400, Request ID: nnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn)" (RequestToken: nnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn, HandlerErrorCode: InvalidRequest)
```

### Reproduction Steps

The following integrity test fails, demonstrating the problem:

```ts
import * as ec2 from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import * as integ from '@aws-cdk/integ-tests';
import * as batch from '../lib/';

export const app = new cdk.App();

const stack = new cdk.Stack(app, 'batch-stack');

const vpc = new ec2.Vpc(stack, 'vpc');

// While this test specifies EFA, the same behavior occurs with
// interfaceType: 'interface' as well
const launchTemplateEFA = new ec2.CfnLaunchTemplate(stack, 'ec2-launch-template-efa', {
  launchTemplateData: {
    networkInterfaces: [{
      deviceIndex: 0,
      subnetId: vpc.privateSubnets[0].subnetId,
      interfaceType: 'efa',
    }],
  },
});

new batch.ComputeEnvironment(stack, 'EFABatch', {
  managed: true,
  computeResources: {
    type: batch.ComputeResourceType.ON_DEMAND,
    instanceTypes: [new ec2.InstanceType('c5n')],
    vpc,
    launchTemplate: {
      launchTemplateName: launchTemplateEFA.launchTemplateName as string,
    },
  },
});

new integ.IntegTest(app, 'BatchWithEFATest', {
  testCases: [stack],
});

app.synth();
```

### Possible Solution

The connected pull request is a proposed solution to this problem, allowing the user to explicitly exclude securitygroups from their ComputeEnvironment, so that they can then set the SecurityGroups in their LaunchTemplate instead.  It's hard to put a validation check in for this, because the user might not have defined the launch template within the stack at all, so its contents cannot be checked before runtime.

### Additional Information/Context

Interestingly, Terraform's AWS provider has exactly the same problem, also recently reported:  https://github.com/hashicorp/terraform-provider-aws/issues/25801

### CDK CLI Version

2.37.1

### Framework Version

_No response_

### Node.js Version

14.20

### OS

MacOS 12.5

### Language

Typescript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by tcutts**
HPC Batch applications frequently require Elastic Fabric Adapters for low-latency networking.  Currently, the `ComputeEnvironment` construct always automatically defines a set of `SecurityGroupIds` in the CloudFormation it generates, and this prevents the stack deploying if the LaunchTemplate contains network interface definitions; Batch does not allow SecurityGroups at the `ComputeEnvironment` level if there are network interfaces defined in the `CfnLaunchTemplate`.

Since we do not currently have support for network interfaces this PR adds a new boolean property in `launchTemplate` called `useNetworkInterfaceSecurityGroups`. When this is enabled we will assume that security groups are being provided by the launch template.

A long term solution may be to:
- Add support for network interfaces in the L2 ec2.LaunchTemplate construct.
- Update the batch.ComputeEnvironment construct to take a ILaunchTemplate instead of the name/id.
- Check the ILaunchTemplate for whether the ComputeEnvironment needs to create any security groups.

closes #21577 

----

### All Submissions:

* [yes] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [no] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [yes] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [yes] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-batch/lib/compute-environment.ts [hunk 1]
```diff
@@ -142,7 +161,7 @@ export interface ComputeResources {
    *
    * @default - AWS default security group.
    */
-  readonly securityGroups?: ec2.ISecurityGroup[];
+  readonly securityGroups?: ec2.ISecurityGroup[] | ComputeEnvironmentSecurityGroups;
```
Comment (corymhall) [inline packages/@aws-cdk/aws-batch/lib/compute-environment.ts:164]: Unfortunately we can't use union types due to limitations with JSII.

**Review by corymhall — CHANGES_REQUESTED** [review]
@tcutts thanks for taking the time to create this PR! Unfortunately the union type is not currently supported in CDK so I don't think this solution will work.

Without looking into this too much, I think the ideal solution is probably a much bigger effort. I'm thinking that we need to:

1. Add support for network interfaces in the L2 ec2.LaunchTemplate construct.
2. Update the batch.ComputeEnvironment construct to take a `ILaunchTemplate` instead of the name/id.
3. Check the `ILaunchTemplate` for whether the `ComputeEnvironment` needs to create any security groups.

**Comment by tcutts** [comment]
> On 15 Aug 2022, at 13:30, Cory Hall ***@***.***> wrote:
> 
> 
> @corymhall requested changes on this pull request.
> 
> @tcutts <https://github.com/tcutts> thanks for taking the time to create this PR! Unfortunately the union type is not currently supported in CDK so I don't think this solution will work.
> 
> 

Ah, I didn’t know that. It did feel a bit dirty.  :-)

> Without looking into this too much, I think the ideal solution is probably a much bigger effort. I'm thinking that we need to:
> 
> Add support for network interfaces in the L2 ec2.LaunchTemplate construct
> Update the batch.ComputeEnvironment construct to take a ILaunchTemplate instead of the name/id.

That would be a breaking change, wouldn’t it?  I suppose the API is currently experimental, so that doesn’t matter quite so much...

> Check the ILaunchTemplate for whether the ComputeEnvironment needs to create any security groups.

… I agree with this, because the combination of all three steps allows us to put a validation in.

**Comment by tcutts** [comment]
This is potentially an enormous can of worms.  CDK has no abstraction for networkInterfaces at all, anywhere.  That needs to be done first before anything else that ought to support them can do so (Instance, AutoScalingGroup).   This is way beyond by relatively hobbyist ability to implement.

**Comment by tcutts** [comment]
What I could do is revert to a simpler change, which is to change what happens if the user specifies `[]` as the list of security groups.  If, in that case, we emit no SecurityGroupIds line in the resource, we still fix the bug, and we've no longer created the unsupported union type.  That does leave us with slightly messy and misleading end user code code:

```ts
const computeEnvironmentEFA = new batch.ComputeEnvironment(stack, 'EFAComputeEnv', {
  managed: true,
  computeResources: {
    securityGroups: [],
    vpc,
    launchTemplate: {
      launchTemplateName: launchTemplateEFA.launchTemplateName as string,
    },
  },
});
```

which I don't like very much, but would achieve the desired outcome.  Arguably, ComputeEnvironment needs updating to do this anyway, because emitting an empty list in a set SecurityGroupIds property doesn't seem to be the right behaviour to me.

**Comment by corymhall** [comment]
> What I could do is revert to a simpler change, which is to change what happens if the user specifies `[]` as the list of security groups. If, in that case, we emit no SecurityGroupIds line in the resource, we still fix the bug, and we've no longer created the unsupported union type. That does leave us with slightly messy and misleading end user code code:

I agree that this isn't what we want, but since this package is experimental and given the complexity of the preferred solution I think I'm fine with this as an interim solution.

File: packages/@aws-cdk/aws-batch/README.md [hunk 2]
```diff
@@ -187,6 +187,42 @@ const myComputeEnv = new batch.ComputeEnvironment(this, 'ComputeEnv', {
...
+      interfaceType: 'efa',
+      groups: [efaSecurityGroup.securityGroupId],
+    }],
+  },
+});
+
+const computeEnvironmentEFA = new batch.ComputeEnvironment(this, 'EFAComputeEnv', {
+  managed: true,
+  computeResources: {
+    useLaunchTemplateNetworkInterface: true,
```
Comment (corymhall) [inline packages/@aws-cdk/aws-batch/README.md:217]: Can you move this property into the `launchTemplate` property? I think it makes sense to provide it as part of the launch template information.

File: packages/@aws-cdk/aws-batch/lib/compute-environment.ts [hunk 3]
```diff
@@ -138,9 +149,11 @@ export interface ComputeResources {
   readonly instanceTypes?: ec2.InstanceType[];
 
   /**
-   * The EC2 security group(s) associated with instances launched in the compute environment.
+   * Up to 5 EC2 security group(s) associated with instances launched in the compute environment.
+   *
+   * This parameter is mutually exclusive with useLaunchTemplateInterface
    *
-   * @default - AWS default security group.
+   * @default - Create a single default security group.
```
Comment (corymhall) [inline packages/@aws-cdk/aws-batch/lib/compute-environment.ts:156]: Good catch!

File: packages/@aws-cdk/aws-batch/lib/compute-environment.ts [hunk 4]
```diff
@@ -520,6 +535,11 @@ export class ComputeEnvironment extends Resource implements IComputeEnvironment,
         if (props.computeResources.spotFleetRole !== undefined) {
           throw new Error('Spot fleet role must not be set for Fargate compute environments');
         }
+
+        // useNetworkInterfaces cannot be specified for Fargate
+        if (props.computeResources.useLaunchTemplateNetworkInterface !== undefined) {
+          throw new Error('LaunchTemplate network interfaces must not be set for Fargate compute environments');
```
Comment (corymhall) [inline packages/@aws-cdk/aws-batch/lib/compute-environment.ts:541]: 👍

**Review by corymhall — CHANGES_REQUESTED** [review]
Looks good! just 1 minor change requested.

**Review by corymhall — APPROVED** [review]

**Pull request merged**